bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/graphics/Image.java

```diff
@@ -820,6 +848,44 @@ long getHandle (int targetZoom, int nativeZoom) {
 	return getImageMetadata(zoomContext).handle;
 }
 
+@FunctionalInterface
+interface ImageDrawFunction {
+    void draw(long imageHandle, Point targetSize);
+}
+
+void executeOnImageHandleAtSizeWithZoomFallback(ImageDrawFunction drawFunction, int targetWidth,
+		int targetHeight) {
+	ImageData imageData;
+	if (this.imageProvider instanceof DynamicallySizableImageDataProvider provider) {
+		imageData = provider.loadImageData(targetWidth, targetHeight);
+	} else {
+		Rectangle bounds = getBounds(100);
+		int imageZoomForWidth = Math.round(100 * targetWidth / bounds.width);
+		int imageZoomForHeight = Math.round(100 * targetHeight / bounds.height);
+		int imageZoom = DPIUtil.getZoomForAutoscaleProperty(Math.max(imageZoomForWidth, imageZoomForHeight));
+		imageData = getImageDataUnscaled(imageZoom);
```

Comment:
Why not have every wrapper implement a `loadImageData(width, height)` method and let them have this implementation in the most abstract class?
Then you can completely remove the additional dynamic wrappers (for input stream, file name provider and data provider) and instead overwrite the `loadImageData(width, height)` methods for the `ImageFileNameProviderWrapper`, `ImageDataProviderWrapper` and `ImageDataLoaderStreamProviderWrapper`, such that they use the dynamic creation logic if their input is capable of doing so. The changes should become much more local then.

Reply:
pushing loadImageData(width, height) into the base abstraction would definitely make the design cleaner and remove the need for instanceof.

The only issue I see is that in the current implementation the “else” branch doesn’t really load an image at the exact requested width/height. Instead, it just picks the closest available zoom level and returns that, which means the resulting ImageData dimensions may not match the loaded size arguments.

So while we could expose a unified loadImageData(targetWidth, targetHeight) method, it might be misleading if callers assume those dimensions are guaranteed. But as the wrappers are private I can just put a comment in the most abstract class the arguments are hints.

Reply:
I’ve moved `loadImageData `into the base abstraction. Instead of calling it directly as `loadImageDataClosestTo`, I’ve renamed the method to `loadImageDataAtSize` and updated the argument names to `widthHint` and `heightHint` to make it clear these are hints, not guaranteed dimensions.

`loadImageDataAtSize` now internally calls `loadImageDataAtExactSize`, which returns the image if an exact match is available. If not, it falls back to `loadImageDataClosestTo` to provide the nearest available size.